Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean

```diff
@@ -0,0 +1,318 @@
+/-
+Copyright (c) 2023 Antoine Chambetr-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambetr-Loir
+-/
+
+import Mathlib.LinearAlgebra.TensorProduct
+import Mathlib.LinearAlgebra.Quotient
+
+/-! # Right exactness properties of tensor product
+
+* For linear maps `f : M →ₗ[R] N`  and `g : N →ₗ[R] P`, `Exact f g` says that `range f = ker g``
+
+* `TensorProduct.rTensor_exact` says that one can tensor a short exact sequence on the right
+
+* `TensorProduct.lTensor_exact` says that one can tensor a short exact sequence on the left
+
+* `TensorProduct.map_ker` computes the kernel of `TensorProduct.map g g'`
+in the presence of two short exact sequences.
+
+The proofs are those of Bourbaki, Algèbre, chap. 2, §3, n°6
+
+## TODO
+
+Add
+
+* Special case for short exact sequences of the form `M → N → N/M → 0`
+
+* Analogue for morphisms of algebras
+
+* Treat the noncommutative case
+
+* Treat the case of semirings (need to know what a short exact sequence is)
+
+-/
+
+
+namespace AddMonoidHom
+
+def Exact {M N P : Type _} [AddCommGroup M] [AddCommGroup N] [AddCommGroup P]
+  (f : outParam (M →+ N)) (g : outParam (N →+ P)) :=
+  g.ker = f.range
+
+end AddMonoidHom
+
+namespace LinearMap
+
+open AddMonoidHom LinearMap
+
+variable (R : Type _) [CommRing R]
+variable {M N P : Type _} [AddCommGroup M] [AddCommGroup N][AddCommGroup P]
+  [Module R M] [Module R N] [Module R P]
+
+variable (f : M →ₗ[R] N) (g : N →ₗ[R] P)
+
+variable {R}
+def Exact := LinearMap.range f = LinearMap.ker g
+
+namespace Exact
+
+variable {f g}
+lemma comp_eq_zero (h : Exact f g) : g.comp f = 0 := by
+  ext x
+  simp only [coe_comp, Function.comp_apply, zero_apply]
+  rw [← mem_ker, ← h]
+  apply mem_range_self
+
+section rTensor
+
+
+open scoped TensorProduct
+
+example (N' : Submodule R N) : N' →ₗ[R] N := by
+  exact Submodule.subtype N'
+
+example (N' : Submodule R N) : Submodule R (N ⊗[R] P) :=
+  LinearMap.range (rTensorHom P (N'.subtype))
+
+
+variable (Q : Type _) [AddCommGroup Q] [Module R Q]
+
+variable (g)
+lemma rTensor.mem_range {p : P} {q : Q} (hp : p ∈ LinearMap.range g) :
+  p ⊗ₜ[R] q ∈  LinearMap.range (rTensor Q g) := by
+  simp only [mem_range] at hp ⊢
+  obtain ⟨n, rfl⟩ := hp
+  use n ⊗ₜ[R] q
+  simp only [rTensor_tmul]
```

Comment:
I think a nicer statement is something like:
```lean
lemma le_comap_range_lTensor (q : Q) :
    LinearMap.range g ≤ (LinearMap.range (lTensor Q g)).comap (TensorProduct.mk R Q P q) := by
  rintro x ⟨n, rfl⟩
  exact ⟨q ⊗ₜ[R] n, rfl⟩

lemma lTensor.mem_range {p : P} (q : Q) (hp : p ∈ LinearMap.range g) :
    q ⊗ₜ[R] p ∈  LinearMap.range (lTensor Q g) := le_comap_range_lTensor g q hp
```
where I'd recommend dropping the second lemma unless you really need it.

Reply:
Done.